### PR TITLE
Bed 17

### PR DESCRIPTION
**Pull request by jwnasambu**

### CODE REVIEW

**Comment by dkayiwa** [comment]
@jwnasambu can you update the ticket with the link to your pull request as advised at? https://openmrs.atlassian.net/wiki/spaces/docs/pages/25477199/Pull+Request+Tips

File: api/src/main/java/org/openmrs/module/bedmanagement/dao/impl/BedManagementDaoImpl.java [hunk 1]
```diff
@@ -129,8 +129,9 @@ public List<BedPatientAssignment> getBedPatientAssignmentByVisit(String visitUui
 	public List<BedPatientAssignment> getCurrentAssignmentsByBed(Bed bed) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/bedmanagement/dao/impl/BedManagementDaoImpl.java]: Can we have the `getCurrentAssignmentsByBed ` method called/not mocked, in a test?

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementServiceImplTest.java [hunk 2]
```diff
@@ -43,30 +44,30 @@
...
 	@Mock
 	BedManagementDao bedManagementDao;
-	
+
 	@Before
 	public void setup() {
 		bedManagementService = new BedManagementServiceImpl();
 		bedManagementService.setDao(bedManagementDao);
 	}
-
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementServiceImplTest.java]: Can we discard these formatting changes?

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementServiceImplTest.java [hunk 3]
```diff
@@ -177,49 +228,50 @@ public void shouldNotCallDeleteBedLocationMappingWhenDeletedBedIsNotMappedToAnyL
 		bed.setStatus(BedStatus.AVAILABLE.name());
 		mockStatic(Context.class);
 		User user = mock(User.class);
-
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementServiceImplTest.java]: It would be great if we discarded these formatting changes.

File: omod/src/main/resources/moduleApplicationContext.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (wikumChamith) [inline omod/src/main/resources/moduleApplicationContext.xml]: Could you let me know why you created this file?

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import org.springframework.test.context.ContextConfiguration;
+
+import java.util.Date;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+@ContextConfiguration(locations = {"classpath:moduleApplicationContext.xml"})
+public class BedManagementDaoImplTest extends BaseModuleContextSensitiveTest {
```
[the remark this answers is not in the document]
Comment (jwnasambu) [reply api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: Thanks @wikumChamith. I now understand why the duplicate ehcache jars/configs on classpath causing the Spring fails to create the sessionFactory. I have deleted the file.

File: omod/src/main/resources/moduleApplicationContext.xml (hunk 4 above)
Comment (jwnasambu) [reply omod/src/main/resources/moduleApplicationContext.xml]: @wikumChamith Thanks for the point out by removing the duplicate and running this command `ls api/target/classes/moduleApplicationContext.xml` gives this output `api/target/classes/moduleApplicationContext.xml`confirming Maven has copied `moduleApplicationContext.xml` into target/classes for the test runtime.

**Comment by jwnasambu** [comment]
@dkayiwa Kindly feel free to review my PR please!

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementServiceImplTest.java [hunk 6]
```diff
@@ -150,6 +152,54 @@ public void shouldGetBedDetailsWithMultiplePatientInformationsById() {
 		assertEquals("M", actualPatient2.getGender());
 	}
 	
+	@Test
+	public void shouldReturnCurrentBedAssignmentWhenVoidedIsFalse() {
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementServiceImplTest.java]: Since this ticket is about queries for data that is already in the database, i find it better to simply include some voided rows in these dataset files and then ensure that they are excluded from the returned results: https://github.com/openmrs/openmrs-module-bedmanagement/tree/master/api/src/test/resources

**Comment by jwnasambu** [comment]
The error is visible now

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    @Before
+    public void setup() throws Exception {
+        executeDataSet("org/openmrs/include/standardTestDataset.xml");
+
+        bedManagementDao = Context.getRegisteredComponent("bedManagementDao", BedManagementDao.class);
+        assertNotNull("bedManagementDao should be initialized", bedManagementDao);
+
+        defaultLocation = Context.getLocationService()
+                .getLocationByUuid("8d6c993e-c2cc-11de-8d13-0010c6dffd0f");
+        assertNotNull("Default location should not be null", defaultLocation);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: I do not think we need this assertion because we are sure that we have this location in the test dataset.

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        bed.setBedNumber(bedNumber);
+        bed.setStatus("AVAILABLE");
+        bed.setBedType(bedType);
+
+        return bedManagementDao.saveBed(bed);
+    }
+
+    private Visit getTestVisit(Patient patient) {
+        List<Visit> visits = Context.getVisitService().getVisitsByPatient(patient);
+        if (visits.isEmpty()) {
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: Do you need to test if visits is empty when you are dealing with visits in a known test dataset file?

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+		Bed bed = new Bed();
+		bed.setBedNumber(bedNumber);
+		bed.setStatus("AVAILABLE");
+		bed.setBedType(bedType);
+		
+		return bedManagementDao.saveBed(bed);
+	}
+	
+	private Visit getTestVisit(Patient patient) {
+		return Context.getVisitService().getVisitsByPatient(patient).get(0);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: You can make this test more efficient by directly fetching this encounter with VisitService.getVisit(id)

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+		Bed bed = new Bed();
+		bed.setBedNumber(bedNumber);
+		bed.setStatus("AVAILABLE");
+		bed.setBedType(bedType);
+		
+		return bedManagementDao.saveBed(bed);
+	}
+	
+	private Visit getTestVisit(Patient patient) {
+		List<Visit> visits = Context.getVisitService().getVisitsByPatient(patient);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: What i mean is Context.getVisitService().getVisit(visitId)

**Comment by jwnasambu** [comment]
@dkayiwa Kindly feel free to review my PR at your convinient time please!

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+		executeDataSet("org/openmrs/include/standardTestDataset.xml");
+		
+		bedManagementDao = Context.getRegisteredComponent("bedManagementDao", BedManagementDao.class);
+		
+		defaultEncounterType = Context.getEncounterService().getEncounterType("Scheduled");
+	}
+	
+	private Bed createBed(String bedNumber) {
+		BedType bedType = new BedType();
+		bedType.setName("Test BedType");
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: Why don't we just load this bed from the test dataset?

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	
+	private Visit getTestVisit(Patient patient, Integer visitId) {
+		Visit visit = Context.getVisitService().getVisit(visitId);
+		assertNotNull(visit);
+		return visit;
+	}
+	
+	private Encounter getTestEncounter(Patient patient, Integer encounterId) {
+		Encounter encounter = Context.getEncounterService().getEncounter(encounterId);
+		assertNotNull(encounter);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: Same as above.

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+	@Test
+	public void shouldReturnBedByPatient() {
+		Patient patient = Context.getPatientService().getPatient(2);
+		Bed bed = createBed("B101");
+		
+		Encounter encounter = getTestEncounter(patient, 3);
+		
+		BedPatientAssignment assignment = new BedPatientAssignment();
+		assignment.setBed(bed);
+		assignment.setPatient(patient);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: Instead of doing all this, just have this data in the test dataset file and then it gets loaded as expected.

File: api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+	public void shouldReturnBedPatientAssignmentByUuid() {
+		Patient patient = Context.getPatientService().getPatient(2);
+		Bed bed = createBed("B102");
+		
+		Encounter encounter = getTestEncounter(patient, 3);
+		
+		BedPatientAssignment assignment = new BedPatientAssignment();
+		assignment.setBed(bed);
+		assignment.setPatient(patient);
+		assignment.setEncounter(encounter);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/bedmanagement/BedManagementDaoImplTest.java]: Same as above.

File: api/src/test/resources/bedManagementDAOComponentTestDataset.xml [hunk 15]
```diff
@@ -171,7 +179,7 @@
 
     <person person_id="567" gender="" birthdate="1975-06-30" birthdate_estimated="false" dead="false" creator="1" date_created="2005-01-01 00:00:00.0" changed_by="1" date_changed="2007-09-20 21:54:12.0" voided="false" void_reason="" uuid="6adb7c42-cfd2-4301-b53b-ff17c5654ff7"/>
     <person person_id="678" gender="" birthdate="1975-06-30" birthdate_estimated="false" dead="false" creator="1" date_created="2005-01-01 00:00:00.0" changed_by="1" date_changed="2007-09-20 21:54:12.0" voided="false" void_reason="" uuid="6adb7c42-cfd2-4301-b53b-ff17c5654dd7"/>
-    
+
     <encounter encounter_id="1001" visit_id="1001" encounter_type="2" patient_id="1001" location_id="1" form_id="1" encounter_datetime="2008-08-01 00:00:00.0" creator="1" date_created="2008-08-18 14:09:05.0" voided="false" uuid="12345678-393b-4118-9c83-a3715b82d4de"/>
```
[the remark this answers is not in the document]
Comment (jwnasambu) [reply api/src/test/resources/bedManagementDAOComponentTestDataset.xml]: Thanks for the point out. Though still fixing the change format issues with the help of Daniel
